### PR TITLE
[V3] WIP JavaScript transformer

### PR DESCRIPTION
**Pull request by alshdavid**
WIP, here for comments and diff

Adding in preliminary support for `Transformer.transform()` in Nodejs plugins

### CODE REVIEW

File: crates/atlaspack_core/src/types/asset.rs [hunk 1]
```diff
@@ -74,6 +76,8 @@ fn create_asset_id(
 pub struct Asset {
   /// The main identify hash for the asset. It is consistent for the entire
   /// build and between builds.
+  #[serde(serialize_with = "se_to_string")]
+  #[serde(deserialize_with = "de_from_str")]
   pub id: u64,
```
Comment (alshdavid) [inline crates/atlaspack_core/src/types/asset.rs:81]: `u64` is not a valid Nodejs type and does not serialize/deserialize correctly. I am transferring it to a string to get around this

File: crates/atlaspack_plugin_rpc/src/nodejs/rpc_conns_nodejs.rs [hunk 2]
```diff
@@ -22,14 +22,13 @@ impl NodejsWorkerFarm {
...
-      *current_index = *current_index + 1;
-    }
-    current_index.clone()
+  fn next_index(&self) -> anyhow::Result<usize> {
+    self
+      .current_index
+      .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |value| {
+        Some((value + 1) % self.workers.len())
+      })
+      .map_err(|v| anyhow::anyhow!(v))
```
Comment (alshdavid) [inline crates/atlaspack_plugin_rpc/src/nodejs/rpc_conns_nodejs.rs:31]: Super basic load balancer that does round-robin to worker threads

File: crates/atlaspack_plugin_rpc/src/plugin/transformer.rs [hunk 3]
```diff
@@ -28,21 +31,29 @@ impl RpcTransformerPlugin {
...
+
+  fn get_key(&self) -> anyhow::Result<&String> {
+    self.key.get_or_try_init(move || {
+      (&self.rpc_worker)
+        .register_transformer(&self.plugin.resolve_from, &self.plugin.package_name)?;
+      Ok(format!(
+        "{}:{}",
+        &self.plugin.resolve_from.to_str().unwrap(),
+        &self.plugin.package_name
+      ))
```
Comment (alshdavid) [inline crates/atlaspack_plugin_rpc/src/plugin/transformer.rs:49]: Initialize the transformer in Js lazily

File: packages/core/core/src/atlaspack-v3/plugins/asset.js [hunk 4]
```diff
@@ -0,0 +1,285 @@
+/* eslint-disable no-unused-vars */
```
Comment (alshdavid) [inline packages/core/core/src/atlaspack-v3/plugins/asset.js:1]: Glue code

File: crates/atlaspack_core/src/serde.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+pub fn de_from_str<'de, D, T>(deserializer: D) -> Result<T, D::Error>
+where
+  T: FromStr,
+  D: Deserializer<'de>,
+{
+  let s = String::deserialize(deserializer)?;
+  let Ok(value) = T::from_str(&s) else {
+    return Err(de::Error::custom(format!("Unable to deserialize {}", s)));
+  };
+  Ok(value) //.map_err(de::Error::custom)
```
Comment (MonicaOlejniczak) [inline crates/atlaspack_core/src/serde.rs:14]:
```suggestion
  Ok(value)
```

File: crates/atlaspack_plugin_rpc/src/nodejs/rpc_conn_nodejs.rs [hunk 6]
```diff
@@ -11,15 +14,20 @@ use crate::RpcWorker;
...
     Ok(Self {
       ping_fn: JsCallable::new_from_object_prop_bound("ping", &delegate)?,
-      transformer_register_fn: JsCallable::new_from_object_prop_bound(
-        "transformerRegister",
+      register_transformer_fn: JsCallable::new_from_object_prop_bound(
+        "registerTransformer",
+        &delegate,
+      )?,
+      transform_transformer_fn: JsCallable::new_from_object_prop_bound(
+        "transformTransformer",
```
Comment (MonicaOlejniczak) [inline crates/atlaspack_plugin_rpc/src/nodejs/rpc_conn_nodejs.rs:30]: /nit but `transform` would be sufficient given it only means one thing in parcel

File: crates/atlaspack_plugin_rpc/src/plugin/transformer.rs [hunk 7]
```diff
@@ -28,21 +31,29 @@ impl RpcTransformerPlugin {
...
+    })
+  }
+
+  fn get_key(&self) -> anyhow::Result<&String> {
+    self.key.get_or_try_init(move || {
+      (&self.rpc_worker)
+        .register_transformer(&self.plugin.resolve_from, &self.plugin.package_name)?;
+      Ok(format!(
+        "{}:{}",
+        &self.plugin.resolve_from.to_str().unwrap(),
```
Comment (MonicaOlejniczak) [inline crates/atlaspack_plugin_rpc/src/plugin/transformer.rs:47]: You should default the unwrap

File: packages/core/core/src/atlaspack-v3/plugins/asset.js [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+import type SourceMap from '@parcel/source-map';
+import {Readable} from 'stream';
+
+export type NapiTransformResult = {|
+  asset: NapiAsset,
+  dependencies: Array<NapiDependency>,
+  invalidateOnFileChange: Array<string>,
+|};
+
+export type NapiAsset = {|
```
Comment (MonicaOlejniczak) [inline packages/core/core/src/atlaspack-v3/plugins/asset.js:27]: A lot of these could be typed can't they?

File: packages/core/core/src/atlaspack-v3/plugins/transformer.js [hunk 9]
```diff
@@ -0,0 +1 @@
+// @flow
```
Comment (MonicaOlejniczak) [inline packages/core/core/src/atlaspack-v3/plugins/transformer.js:1]: Delete this file?

File: crates/atlaspack_core/src/lib.rs [hunk 10]
```diff
@@ -4,4 +4,5 @@ pub mod cache;
 pub mod config_loader;
 pub mod hash;
 pub mod plugin;
+pub mod serde;
```
Comment (MonicaOlejniczak) [inline crates/atlaspack_core/src/lib.rs:7]: Can be private

**Pull request closed without merging**